### PR TITLE
test: add OpenAI telemetry utility tests

### PR DESCRIPTION
**Pull request by koxudaxi**
### TL;DR

Added unit tests for OpenAI telemetry utilities in the Lilypad SDK.

### What changed?

This PR adds comprehensive unit tests for the OpenAI telemetry utilities in the Lilypad SDK. The tests cover:

- `StreamWrapper` and `AsyncStreamWrapper` classes for handling both synchronous and asynchronous OpenAI API response streams
- `OpenAIMetadata` and `OpenAIChunkHandler` for extracting and processing metadata from API responses
- Utility functions like `default_openai_cleanup`, `set_message_event`, `set_response_attributes`, `get_tool_calls`, and `get_choice_event`

The tests verify proper handling of various scenarios including:
- Normal streaming responses
- Error handling
- Context manager behavior
- Tool calls processing
- Different message formats and content types

### How to test?

Run the unit tests with pytest:

```bash
cd sdks/python
pytest tests/lilypad/_internal/otel/openai/test_utils.py -v
```

### Why make this change?

These tests ensure the reliability and correctness of the OpenAI telemetry utilities, which are critical for proper observability when using OpenAI's APIs through the Lilypad SDK. The tests cover edge cases and various response formats, helping to prevent regressions and ensure consistent behavior across different OpenAI API interactions.

### CODE REVIEW

**Comment by koxudaxi** [comment]
* **#655** <a href="https://app.graphite.dev/github/pr/Mirascope/lilypad/655?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Mirascope/lilypad/655?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#654** <a href="https://app.graphite.dev/github/pr/Mirascope/lilypad/654?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#653** <a href="https://app.graphite.dev/github/pr/Mirascope/lilypad/653?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#652** <a href="https://app.graphite.dev/github/pr/Mirascope/lilypad/652?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#651** <a href="https://app.graphite.dev/github/pr/Mirascope/lilypad/651?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#647** <a href="https://app.graphite.dev/github/pr/Mirascope/lilypad/647?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#646** <a href="https://app.graphite.dev/github/pr/Mirascope/lilypad/646?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `v1`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by willbakst — COMMENTED** [review]
Same thing here re: mocking. I will revisit with a true review once we sync on what the best approach to testing is here.

**Comment by willbakst** [comment]
I think we should do the same thing here with `_patch` and make this `_utils` and not unit test these and instead get coverage through `test_instrument` using `vcr.py` and `inline-snapshot`.

The core idea is that we are testing the public `instrument_openai` interface while the `_patch` and `_utils` methods are private implementation details that can change without breaking our tests.

File: sdks/python/tests/lilypad/_internal/otel/openai/test_instrument.py
Comment (willbakst) [inline sdks/python/tests/lilypad/_internal/otel/openai/test_instrument.py:1]: Not sure if these changes make sense given down stack updates? Do we need these tests?

File: sdks/python/tests/lilypad/_internal/otel/openai/test_patch.py
Comment (willbakst) [inline sdks/python/tests/lilypad/_internal/otel/openai/test_patch.py:1]: Same thing here re: do we need these tests?

File: sdks/python/tests/lilypad/_internal/otel/openai/test_utils.py
Comment (willbakst) [inline sdks/python/tests/lilypad/_internal/otel/openai/test_utils.py:1]: I'm wondering if we should do the same thing here and make the module `_utils` and use it the same way we do `_patch` and then only test `instrument`?

I feel like these are private interfaces and again we're testing our implementation rather than functionality.

File: sdks/python/src/lilypad/_internal/otel/openai/_utils.py [hunk 1]
```diff
@@ -178,23 +176,18 @@ def get_tool_calls(
     return calls
 
 
-def set_message_event(
-    span: Span, message: ChatCompletionMessage | ChatCompletionMessageParam
-) -> None:
+def set_message_event(span: Span, message: ChatCompletionMessageParam) -> None:
```
Comment (willbakst) [inline sdks/python/src/lilypad/_internal/otel/openai/_utils.py:178]: Can you confirm that we only use this method for input messages? Do we not set message events for output messages?

File: sdks/python/tests/lilypad/_internal/otel/openai/test_instrument.py [hunk 2]
```diff
@@ -627,53 +977,139 @@ def test_multiple_clients_independent(api_key: str):
...
 
 @pytest.mark.vcr()
-def test_error_handling(client: OpenAI, span_exporter: InMemorySpanExporter):
-    """Test error handling and span error recording."""
-    with pytest.raises(NotFoundError):
-        client.chat.completions.create(
-            model="invalid-model-name",
-            messages=[{"role": "user", "content": "Hello"}],
-        )
+def test_tool_message_with_tool_call_id(
```
Comment (willbakst) [inline sdks/python/tests/lilypad/_internal/otel/openai/test_instrument.py:748]: We should sync on which tests need only sync and which need both sync and async so we can reduce the required tests only to those that are necessary

File: sdks/python/tests/lilypad/_internal/otel/openai/test_instrument.py [hunk 3]
```diff
@@ -733,3 +1169,743 @@ class TestResponse(BaseModel):
...
+                        "finish_reason": "stop",
+                        "message": '{"role": "assistant", "content": "Goodbye! If you ever need anything in the future, feel free to reach out. Take care!"}',
+                    },
+                },
+            ],
+        }
+    )
+
+
+def test_sync_streaming_tool_calls_gap_handles_none(
```
Comment (willbakst) [inline sdks/python/tests/lilypad/_internal/otel/openai/test_instrument.py]: What do you mean by "gap handling"?

File: sdks/python/src/lilypad/_internal/otel/openai/_utils.py (hunk 1 above)
Comment (willbakst) [inline sdks/python/src/lilypad/_internal/otel/openai/_utils.py:178]: nit: given this is only `ChatCompletionMessageParam`, we can do this:

```python
def set_message_event(span: Span, message: ChatCompletionMessageParam) -> None:
    """Add a message event to the span with appropriate attributes."""
    attributes = {
        gen_ai_attributes.GEN_AI_SYSTEM: gen_ai_attributes.GenAiSystemValues.OPENAI.value
    }
    role = message["role"]

    if role == "assistant" and (tool_calls := get_tool_calls(message)):
        attributes["tool_calls"] = json_dumps(tool_calls)

    if role == "tool" and (tool_call_id := message.get("tool_call_id")):
        attributes["id"] = tool_call_id

    if content := message.get("content"):
        if not isinstance(content, str):
            content = json_dumps(content)
        attributes["content"] = content

    span.add_event(
        f"gen_ai.{role}.message",
        attributes=attributes,
    )
```

**Review by willbakst — APPROVED** [review]

**Comment by willbakst** [comment]
### Merge activity

* **Aug 13, 12:28 AM UTC**: @willbakst merged this pull request with [Graphite](https://app.graphite.dev/github/pr/Mirascope/lilypad/655).
